File: extensions/reddit/src/main/java/app/morphe/extension/reddit/patches/ShowViewCountPatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */
package app.morphe.extension.reddit.patches;

import app.morphe.extension.reddit.settings.Settings;
import app.morphe.extension.shared.Logger;
import app.morphe.extension.shared.Utils;

@SuppressWarnings("unused")
public final class ShowViewCountPatch {
    private static final String ANDROID_POST_UNIT_VIEWS_COUNT = "android_post_unit_views_count_v2";

    /**
     * @return If this patch was included during patching.
     */
    public static boolean isPatchIncluded() {
        return false;  // Modified during patching.
    }

    /**
     * Injection point.
     */
    public static boolean showViewCount(String experimentName, boolean original) {
        // Can be called before app context is set.
        if (Utils.getContext() == null) {
            Logger.printInfo(() -> "Cannot show view count, context is not yet set");
            return original;
        }

        if (Settings.SHOW_VIEW_COUNT.get() && experimentName != null
                && experimentName.startsWith(ANDROID_POST_UNIT_VIEWS_COUNT)) {
            return true;
        }

        return original;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/reddit/layout/viewcount/ShowViewCountPatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */
package app.morphe.patches.reddit.layout.viewcount

import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.reddit.misc.flag.featureFlagHookPatch
import app.morphe.patches.reddit.misc.flag.hookFeatureFlag
import app.morphe.patches.reddit.misc.settings.settingsPatch
import app.morphe.patches.reddit.shared.Constants.COMPATIBILITY_REDDIT
import app.morphe.util.setExtensionIsPatchIncluded

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/reddit/patches/ShowViewCountPatch;"

@Suppress("unused")
val showViewCountPatch = bytecodePatch(
    name = "Show view count",
    description = "Adds an option to show the view count of Posts."
) {
    compatibleWith(COMPATIBILITY_REDDIT)

    dependsOn(
        settingsPatch,
        featureFlagHookPatch
    )

    execute {

        hookFeatureFlag("$EXTENSION_CLASS->showViewCount")

        setExtensionIsPatchIncluded(EXTENSION_CLASS)
    }
}
